### ISSUE TITLE
update 4.2.0 to 5.0.4

### ISSUE THREAD

**Issue by michelmelo** [crater-invoice/crater#632]
**Describe the bug**
message: "SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 

![migrate](https://user-images.githubusercontent.com/159223/145045584-e0397f55-4ecc-4596-ba66-ac3c64f7ec09.png)

**Comment by SteveEdson** [comment]
I also had some issues with failed migrations, saying that the `addresses()` method didn't exist on the `User` model

**Comment by michelmelo** [comment]
my case is duplicate users

**Comment by azaricstefan** [comment]
This command fails:

```
php artisan migrate
Migrating: 2021_06_28_105334_create_bouncer_tables

   Illuminate\Database\QueryException 

  SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (SQL: drop table `roles`)

  at vendor/laravel/framework/src/Illuminate/Database/Connection.php:703
    699▕         // If an exception occurs when attempting to run a query, we'll format the error
    700▕         // message to include the bindings with SQL, which will make this exception a
    701▕         // lot more helpful to the developer instead of just the database's errors.
    702▕         catch (Exception $e) {
  ➜ 703▕             throw new QueryException(
    704▕                 $query, $this->prepareBindings($bindings), $e
    705▕             );
    706▕         }
    707▕     }

      +12 vendor frames 
  13  database/migrations/2021_06_28_105334_create_bouncer_tables.php:19
      Illuminate\Support\Facades\Facade::__callStatic("drop")

      +21 vendor frames 
  35  artisan:37
      Illuminate\Foundation\Console\Kernel::handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
```

After manually deleting tables to unblock the migration (roles,permissions,model_has_permissions, model_has_roles and role_has_permissions) I get this problem.

![image](https://user-images.githubusercontent.com/9116773/145120770-f06d6822-9ce2-4cb1-90bc-b1741727f42a.png)

**Comment by azaricstefan** [comment]
If you look into migration `2021_06_30_062411_update_customer_id_in_all_tables.php`

This part of code:
```
                if ($user->addresses()->exists()) {
                    $user->addresses->map(function ($address) use ($newCustomer) {
                        if ($address) {
                            $address->customer_id = $newCustomer->id;
                            $address->user_id = null;
                            $address->save();
                        }
                    });
                }
```
Relation addresses doesn't exist. What's the proper definition for this relation?

In User.php model we could add this:

```
    public function addresses()
    {
        return $this->hasMany(Address::class, 'user_id');
    }
```
is `user_id` correct key here? I don't see what else could be.

**Comment by barf** [comment]
Also reproducing this upgrading 4.2.0 to version 5.
```
SQLSTATE[2BP01]: Dependent objects still exist: 7 ERROR:  cannot drop table roles because other objects depend on it
DETAIL:  constraint model_has_roles_role_id_foreign on table model_has_roles depends on table roles
constraint role_has_permissions_role_id_foreign on table role_has_permissions depends on table roles
HINT:  Use DROP ... CASCADE to drop the dependent objects too. (SQL: drop table "roles")
```

**Comment by mohitpanjwani** [comment]
Hey everyone, 

Please hold off updating until tonight. We're releasing another change that would allow you to run `php artisan migrate` directly on latest code and it also fixes the errors regarding roles & user address relationships. 

--------------
You can stop reading further however, if you'd like to understand what went wrong: 
1. The roles / permissions table was added by a package which we never used in an older version (while testing we found out and dropped these tables if they existed but the order of drop caused the above foreign key error)
2. The customers used to be saved on the Users table before v5 but we updated the relationship to use `customers` table instead. This change unfortunately caused an issue if you directly pulled in the latest git version and ran the migration command ( This issue won't happen if you update using the web ui or update command)

**Comment by mohitpanjwani** [comment]
Hi All,

We've just fixed all the issues with migration on 5.0.5

You can just do a `git pull` with the latest code of Crater and run `php artisan migrate`. That should fix everything. 

If you don't use git, then you can just copy all the latest files from Crater 5.0.5 zip file.

-------------------

You can also update from Settings -> Update App again if you have the old version backup.

**Comment by mohitpanjwani** [comment]
I’m closing this issue but please feel free to re-open if you get any more errors on update. 

Also if you’re stuck on the old update because of any of the above mentioned errors, please send a dm on discord. I’ll help you fix it personally 👍
